topicExtractor.py

```diff
@@ -126,6 +122,12 @@ def initializeTopicModel(self, vectorizerModel=None):
                 representation_model=self.representationModel,
                 vectorizer_model=vectorizerModel,
             )
+        if clusterModel is not None:
+            self.topicModel = BERTopic(
+                representation_model=self.representationModel,
+                vectorizer_model=vectorizerModel,
+                hdbscan_model=clusterModel,
+            )
```

Comment:
Can you explain vectorizing vs Cluster modeling here? I understand this is done due to fact that not enough topic are generated just by vectorizing hence the clustering steps. But How does it help?

Reply:
Vectorizing and Clustering are separate steps. We always use Vectorzing (this is where KeyBERT is used) because vectorizing data is the first step to allow for ingestion into any Deep Learning Model, be it BERT, GPT, anything.  So we always need to vectorize the data. We use KeyBERT here to vectorize the text in such a way that keywords can be focused more on than other filler words that might not be as important. 

(A note: Across BERTopic, we actually perform 5 distinct steps: we first embed the data, then we cluster the data, then vectorize it, and finally apply TF-IDF to it. I can explain in detail further if needed.)

Clustering is a step before vectorization. Now, the function call in BERTopic is a bit confusingly setup, as they even mention in their [documentation here](https://maartengr.github.io/BERTopic/getting_started/clustering/clustering.html):
> As you might have noticed, the cluster_model is passed to hdbscan_model which might be a bit confusing considering you are not passing an HDBSCAN model. For now, the name of the parameter is kept the same to adhere to the current state of the API. Changing the name could lead to deprecation issues, which I want to prevent as much as possible.

So, default clustering uses a technique called HDBSCAN, which is really powerful. The reason why it sometimes fails across certain videos lies in the way it tries to make clusters of topics. Topics are made up of sub-topics, or keywords that are being talked about in that topic. What happens is that HDBSCAN can sometimes be unable to group all of the sub-topics together under one singular topic. This might seem like a problem, but I would argue this is a strength of this approach, it is a quantitative way to determine the concentration and spread of topics in a video. 

When no single main topic is possible for a given video, it can also indicate that a lot of things are being covered in a way where asking a single question for an overarching topic might not even make sense. But I digress. We can still switch to other clustering algorithms, and in this case K-Means. To think of it simply, HDBSCAN has the ability to say some points are outliers. K-Means does not, and hence will try to fit all the data into a specific number of categories, no matter what. This is not ideal, as we have to hardcode the number of categories we want, and there is no way to remove "noisy" data segments. But it still works, and it will cluster the data in a way that LangChain can still look at to determine a potential topic title based on clustered sub-topics found.  It is more like saying that some algorithms can have utility when other algorithms fail, and hence no algorithm is truly inferior to the other one. K-means works, and because it does because HDBSCAN couldn't, it is a good fallback option.

I also would like to mention that these algorithms in itself do not directly affect the question generation steps later in the pipeline. The Question Generator is separate, and will always make a question based on only the text it sees, so the relevance of a question will never be affected by changes we make during topic extraction directly. What changes in BERTopic do affect is the "quality" and "utility" of the text being used to generate questions. One of the goals for the project is to not just generate questions at the right time but to generate usable questions that assist a viewer in learning something. Having this set up for BERTopic in the first stage is what lets us have a level of explainability and tunability in the actual topics of interest being tracked across a video, and why we can even implement a different clustering algorithm when one doesn't work as we expected it to. 

(Sorry, this comment got a lot larger than I expected. I will elaborate further on any aspects here that felt confusing or underexplained if needed.)

Reply:
Thanks @takposha For details, I appreciate your effort.

Reply:
No problem. I will be working on adding a simpler way to at least explain some of this in a presentation or something similar to make this more accessible.